#!/usr/bin/env python3
"""
–ü–æ–ª–Ω—ã–π E2E —Ç–µ—Å—Ç –ø–∞–π–ø–ª–∞–π–Ω–∞: gRPC –∑–∞–ø—Ä–æ—Å ‚Üí LLM –æ—Ç–≤–µ—Ç ‚Üí Edge TTS ‚Üí gRPC –æ—Ç–≤–µ—Ç –∫–ª–∏–µ–Ω—Ç—É

–ü—Ä–æ–≤–µ—Ä—è–µ—Ç –≤–µ—Å—å –ø—Ä–æ—Ü–µ—Å—Å –æ—Ç –º–æ–º–µ–Ω—Ç–∞ –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏ –æ—Ç–≤–µ—Ç–∞ –∞—Å—Å–∏—Å—Ç–µ–Ω—Ç–∞ –¥–æ –ø–µ—Ä–µ–¥–∞—á–∏ –∞—É–¥–∏–æ –Ω–∞ –∫–ª–∏–µ–Ω—Ç—Å–∫—É—é —á–∞—Å—Ç—å:
1. gRPC –∑–∞–ø—Ä–æ—Å –æ—Ç –∫–ª–∏–µ–Ω—Ç–∞
2. –û–±—Ä–∞–±–æ—Ç–∫–∞ —á–µ—Ä–µ–∑ GrpcServiceIntegration
3. StreamingWorkflowIntegration (LLM ‚Üí —Ç–µ–∫—Å—Ç)
4. Edge TTS (—Ç–µ–∫—Å—Ç ‚Üí –∞—É–¥–∏–æ –≤ —Ñ–æ—Ä–º–∞—Ç–µ Azure TTS)
5. –ü–µ—Ä–µ–¥–∞—á–∞ —á–µ—Ä–µ–∑ gRPC –æ–±—Ä–∞—Ç–Ω–æ –∫–ª–∏–µ–Ω—Ç—É
"""

import asyncio
import sys
import logging
from pathlib import Path
from typing import Dict, Any, AsyncGenerator, List
from unittest.mock import Mock, AsyncMock, patch

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ —Å–µ—Ä–≤–µ—Ä—É
server_root = Path(__file__).parent.parent
sys.path.insert(0, str(server_root))

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class MockGrpcContext:
    """–ú–æ–∫ gRPC –∫–æ–Ω—Ç–µ–∫—Å—Ç–∞"""
    def __init__(self):
        self.code = None
        self.details = None
    
    def set_code(self, code):
        self.code = code
    
    def set_details(self, details):
        self.details = details


class MockTextProcessor:
    """–ú–æ–∫ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞ (LLM)"""
    
    def __init__(self):
        self.is_initialized = True
        self.name = "text_processing"
        self.call_count = 0
    
    async def process(self, request: Dict[str, Any]) -> AsyncGenerator[str, None]:
        """–ò–º–∏—Ç–∞—Ü–∏—è –ø–æ—Ç–æ–∫–æ–≤–æ–≥–æ –æ—Ç–≤–µ—Ç–∞ LLM"""
        self.call_count += 1
        text = request.get('text', '')
        
        # –ò–º–∏—Ç–∏—Ä—É–µ–º –ø–æ—Ç–æ–∫–æ–≤—ã–π –æ—Ç–≤–µ—Ç –∞—Å—Å–∏—Å—Ç–µ–Ω—Ç–∞
        responses = [
            "Hello! ",
            "This is a test response from the assistant. ",
            "The audio generation should work correctly. ",
            "Let's verify that the full pipeline works as expected."
        ]
        
        for response in responses:
            await asyncio.sleep(0.05)  # –ò–º–∏—Ç–∞—Ü–∏—è –∑–∞–¥–µ—Ä–∂–∫–∏
            yield response
        
        # –§–∏–Ω–∞–ª—å–Ω—ã–π –ø–æ–ª–Ω—ã–π –æ—Ç–≤–µ—Ç
        await asyncio.sleep(0.05)
        yield ' '.join(responses)


async def test_full_pipeline_e2e():
    """
    –ü–æ–ª–Ω—ã–π E2E —Ç–µ—Å—Ç: gRPC ‚Üí LLM ‚Üí Edge TTS ‚Üí gRPC –∫–ª–∏–µ–Ω—Ç—É
    """
    print("\n" + "="*80)
    print("–ü–û–õ–ù–´–ô E2E –¢–ï–°–¢ –ü–ê–ô–ü–õ–ê–ô–ù–ê")
    print("="*80)
    print("\n–ü—Ä–æ–≤–µ—Ä—è–µ–º –≤–µ—Å—å –ø—Ä–æ—Ü–µ—Å—Å:")
    print("1. gRPC –∑–∞–ø—Ä–æ—Å –æ—Ç –∫–ª–∏–µ–Ω—Ç–∞")
    print("2. –û–±—Ä–∞–±–æ—Ç–∫–∞ —á–µ—Ä–µ–∑ GrpcServiceIntegration")
    print("3. StreamingWorkflowIntegration (LLM ‚Üí —Ç–µ–∫—Å—Ç)")
    print("4. Edge TTS (—Ç–µ–∫—Å—Ç ‚Üí –∞—É–¥–∏–æ –≤ —Ñ–æ—Ä–º–∞—Ç–µ Azure TTS)")
    print("5. –ü–µ—Ä–µ–¥–∞—á–∞ —á–µ—Ä–µ–∑ gRPC –æ–±—Ä–∞—Ç–Ω–æ –∫–ª–∏–µ–Ω—Ç—É")
    print("="*80)
    
    try:
        # –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –º–æ–¥—É–ª–∏
        from modules.audio_generation.core.audio_processor import AudioProcessor
        from integrations.workflow_integrations.streaming_workflow_integration import StreamingWorkflowIntegration
        from integrations.service_integrations.grpc_service_integration import GrpcServiceIntegration
        from modules.grpc_service.core.grpc_server import streaming_pb2
        
        # 1. –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è AudioProcessor (Edge TTS)
        print("\n[1/5] –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è AudioProcessor (Edge TTS)...")
        audio_processor = AudioProcessor()
        audio_init_result = await audio_processor.initialize()
        if not audio_init_result:
            print("   ‚ùå –û—à–∏–±–∫–∞ –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–∏ AudioProcessor")
            return False
        print("   ‚úÖ AudioProcessor –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Ñ–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ
        audio_info = audio_processor.get_audio_info()
        print(f"   üìä –§–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ:")
        print(f"      - Sample rate: {audio_info.get('sample_rate')} Hz")
        print(f"      - Channels: {audio_info.get('channels')}")
        print(f"      - Bits per sample: {audio_info.get('bits_per_sample')}")
        print(f"      - Format: {audio_info.get('format')}")
        print(f"      - Voice: {audio_info.get('voice_name')}")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º –±–∞–∑–æ–≤—ã–µ –ø–∞—Ä–∞–º–µ—Ç—Ä—ã —Ñ–æ—Ä–º–∞—Ç–∞ (—Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç—å —Å Azure TTS)
        # Azure TTS –∏—Å–ø–æ–ª—å–∑—É–µ—Ç: 24kHz, 16-bit, mono, PCM
        # Edge TTS –º–æ–∂–µ—Ç –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –¥—Ä—É–≥–∏–µ sample rates, –Ω–æ —Ñ–æ—Ä–º–∞—Ç PCM –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å —Å–æ–≤–º–µ—Å—Ç–∏–º
        assert audio_info.get('channels') == 1, "Channels –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å 1 (mono, –∫–∞–∫ –≤ Azure TTS)"
        assert audio_info.get('bits_per_sample') == 16, "Bits per sample –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å 16 (–∫–∞–∫ –≤ Azure TTS)"
        assert audio_info.get('format') in ['pcm', 'int16'], "Format –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å PCM –∏–ª–∏ int16 (–∫–∞–∫ –≤ Azure TTS)"
        print("   ‚úÖ –§–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS (PCM, 16-bit, mono)")
        print(f"   ‚ÑπÔ∏è  Sample rate: {audio_info.get('sample_rate')} Hz (–º–æ–∂–µ—Ç –æ—Ç–ª–∏—á–∞—Ç—å—Å—è –æ—Ç Azure TTS 24kHz)")
        
        # 2. –°–æ–∑–¥–∞–Ω–∏–µ –º–æ–∫–∞ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞ (LLM)
        print("\n[2/5] –°–æ–∑–¥–∞–Ω–∏–µ –º–æ–∫–∞ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞ (LLM)...")
        text_processor = MockTextProcessor()
        print("   ‚úÖ –ú–æ–∫ —Ç–µ–∫—Å—Ç–æ–≤–æ–≥–æ –ø—Ä–æ—Ü–µ—Å—Å–æ—Ä–∞ —Å–æ–∑–¥–∞–Ω")
        
        # 3. –°–æ–∑–¥–∞–Ω–∏–µ StreamingWorkflowIntegration
        print("\n[3/5] –°–æ–∑–¥–∞–Ω–∏–µ StreamingWorkflowIntegration...")
        workflow = StreamingWorkflowIntegration(
            text_processor=text_processor,
            audio_processor=audio_processor,
            memory_workflow=None,
            text_filter_manager=None
        )
        await workflow.initialize()
        print("   ‚úÖ StreamingWorkflowIntegration –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        # 4. –°–æ–∑–¥–∞–Ω–∏–µ GrpcServiceIntegration
        print("\n[4/5] –°–æ–∑–¥–∞–Ω–∏–µ GrpcServiceIntegration...")
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å–∏–≥–Ω–∞—Ç—É—Ä—É –∫–æ–Ω—Å—Ç—Ä—É–∫—Ç–æ—Ä–∞
        import inspect
        sig = inspect.signature(GrpcServiceIntegration.__init__)
        params = list(sig.parameters.keys())[1:]  # –ü—Ä–æ–ø—É—Å–∫–∞–µ–º 'self'
        
        # –°–æ–∑–¥–∞–µ–º —Å –ø—Ä–∞–≤–∏–ª—å–Ω—ã–º–∏ –ø–∞—Ä–∞–º–µ—Ç—Ä–∞–º–∏
        init_kwargs = {
            'streaming_workflow': workflow,
        }
        if 'memory_workflow' in params:
            init_kwargs['memory_workflow'] = None
        if 'text_filter_manager' in params:
            init_kwargs['text_filter_manager'] = None
        
        grpc_integration = GrpcServiceIntegration(**init_kwargs)
        await grpc_integration.initialize()
        print("   ‚úÖ GrpcServiceIntegration –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        # 5. –¢–µ—Å—Ç–æ–≤—ã–π –∑–∞–ø—Ä–æ—Å (–∏–º–∏—Ç–∞—Ü–∏—è gRPC –∑–∞–ø—Ä–æ—Å–∞)
        print("\n[5/5] –û–±—Ä–∞–±–æ—Ç–∫–∞ —Ç–µ—Å—Ç–æ–≤–æ–≥–æ –∑–∞–ø—Ä–æ—Å–∞...")
        request_data = {
            'text': 'Hello, can you tell me a test response?',
            'hardware_id': 'test_hardware_e2e',
            'session_id': 'test_session_e2e'
        }
        
        print(f"   üìù –ó–∞–ø—Ä–æ—Å: '{request_data['text']}'")
        
        # –°–æ–±–∏—Ä–∞–µ–º —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫—É
        text_chunks_received = []
        audio_chunks_received = []
        total_audio_bytes = 0
        grpc_responses = []
        
        # –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –∑–∞–ø—Ä–æ—Å —á–µ—Ä–µ–∑ GrpcServiceIntegration
        async for result in grpc_integration.process_request_complete(request_data):
            # –§–æ—Ä–º–∏—Ä—É–µ–º gRPC –æ—Ç–≤–µ—Ç (–∫–∞–∫ –≤ grpc_server.py)
            if 'text_response' in result and result.get('text_response'):
                text = result['text_response']
                text_chunks_received.append(text)
                print(f"   üìù –¢–µ–∫—Å—Ç —á–∞–Ω–∫: '{text[:60]}{'...' if len(text) > 60 else ''}'")
                
                # –ò–º–∏—Ç–∏—Ä—É–µ–º gRPC –æ—Ç–≤–µ—Ç
                grpc_response = streaming_pb2.StreamResponse(text_chunk=text)
                grpc_responses.append(('text', text))
            
            # –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –∞—É–¥–∏–æ —á–∞–Ω–∫–∏
            if 'audio_chunk' in result:
                audio_chunk = result['audio_chunk']
                if isinstance(audio_chunk, (bytes, bytearray)) and len(audio_chunk) > 0:
                    audio_chunks_received.append(audio_chunk)
                    total_audio_bytes += len(audio_chunk)
                    
                    # –ò–º–∏—Ç–∏—Ä—É–µ–º gRPC –æ—Ç–≤–µ—Ç (–∫–∞–∫ –≤ grpc_server.py)
                    grpc_response = streaming_pb2.StreamResponse(
                        audio_chunk=streaming_pb2.AudioChunk(
                            audio_data=audio_chunk,
                            dtype='int16',  # PCM —Ñ–æ—Ä–º–∞—Ç (24kHz, 16-bit, mono)
                            shape=[]
                        )
                    )
                    grpc_responses.append(('audio', len(audio_chunk)))
                    print(f"   üîä –ê—É–¥–∏–æ —á–∞–Ω–∫: {len(audio_chunk)} –±–∞–π—Ç (PCM, int16)")
            
            # –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –º–Ω–æ–∂–µ—Å—Ç–≤–µ–Ω–Ω—ã–µ –∞—É–¥–∏–æ —á–∞–Ω–∫–∏
            if 'audio_chunks' in result:
                for chunk_data in result.get('audio_chunks', []):
                    if chunk_data and len(chunk_data) > 0:
                        audio_chunks_received.append(chunk_data)
                        total_audio_bytes += len(chunk_data)
                        
                        grpc_response = streaming_pb2.StreamResponse(
                            audio_chunk=streaming_pb2.AudioChunk(
                                audio_data=chunk_data,
                                dtype='int16',
                                shape=[]
                            )
                        )
                        grpc_responses.append(('audio', len(chunk_data)))
                        print(f"   üîä –ê—É–¥–∏–æ —á–∞–Ω–∫ (–∏–∑ —Å–ø–∏—Å–∫–∞): {len(chunk_data)} –±–∞–π—Ç")
            
            if result.get('is_final'):
                print(f"   ‚úÖ –§–∏–Ω–∞–ª—å–Ω—ã–π –æ—Ç–≤–µ—Ç –ø–æ–ª—É—á–µ–Ω")
        
        # 6. –ü—Ä–æ–≤–µ—Ä–∫–∞ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
        print("\n" + "="*80)
        print("–†–ï–ó–£–õ–¨–¢–ê–¢–´ –¢–ï–°–¢–ò–†–û–í–ê–ù–ò–Ø")
        print("="*80)
        
        print(f"\nüìä –°—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞:")
        print(f"   - –¢–µ–∫—Å—Ç–æ–≤—ã—Ö —á–∞–Ω–∫–æ–≤: {len(text_chunks_received)}")
        print(f"   - –ê—É–¥–∏–æ —á–∞–Ω–∫–æ–≤: {len(audio_chunks_received)}")
        print(f"   - –í—Å–µ–≥–æ –∞—É–¥–∏–æ –±–∞–π—Ç: {total_audio_bytes} ({total_audio_bytes / 1024:.2f} KB)")
        print(f"   - gRPC –æ—Ç–≤–µ—Ç–æ–≤: {len(grpc_responses)}")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Ñ–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ
        if audio_chunks_received:
            print(f"\nüîç –ü—Ä–æ–≤–µ—Ä–∫–∞ —Ñ–æ—Ä–º–∞—Ç–∞ –∞—É–¥–∏–æ:")
            first_chunk = audio_chunks_received[0]
            print(f"   - –†–∞–∑–º–µ—Ä –ø–µ—Ä–≤–æ–≥–æ —á–∞–Ω–∫–∞: {len(first_chunk)} –±–∞–π—Ç")
            print(f"   - –¢–∏–ø –¥–∞–Ω–Ω—ã—Ö: bytes")
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ —ç—Ç–æ PCM –¥–∞–Ω–Ω—ã–µ (–¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –∫—Ä–∞—Ç–Ω—ã 2 –¥–ª—è 16-bit)
            assert len(first_chunk) % 2 == 0, "PCM 16-bit –¥–∞–Ω–Ω—ã–µ –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –∫—Ä–∞—Ç–Ω—ã 2 –±–∞–π—Ç–∞–º"
            print(f"   ‚úÖ –†–∞–∑–º–µ—Ä —á–∞–Ω–∫–∞ –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω –¥–ª—è 16-bit PCM")
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ —Ñ–æ—Ä–º–∞—Ç —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS
            # Azure TTS –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç raw PCM –±–µ–∑ –∑–∞–≥–æ–ª–æ–≤–∫–æ–≤
            # Edge TTS —Ç–æ–∂–µ –¥–æ–ª–∂–µ–Ω –≤–æ–∑–≤—Ä–∞—â–∞—Ç—å raw PCM –±–µ–∑ –∑–∞–≥–æ–ª–æ–≤–∫–æ–≤
            print(f"   ‚úÖ –§–æ—Ä–º–∞—Ç —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS (raw PCM, –±–µ–∑ –∑–∞–≥–æ–ª–æ–≤–∫–æ–≤)")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º gRPC —Ñ–æ—Ä–º–∞—Ç
        print(f"\nüîç –ü—Ä–æ–≤–µ—Ä–∫–∞ gRPC —Ñ–æ—Ä–º–∞—Ç–∞:")
        audio_grpc_responses = [r for r in grpc_responses if r[0] == 'audio']
        if audio_grpc_responses:
            print(f"   - –ê—É–¥–∏–æ gRPC –æ—Ç–≤–µ—Ç–æ–≤: {len(audio_grpc_responses)}")
            print(f"   - –í—Å–µ –∏—Å–ø–æ–ª—å–∑—É—é—Ç dtype='int16' (PCM —Ñ–æ—Ä–º–∞—Ç)")
            print(f"   ‚úÖ gRPC —Ñ–æ—Ä–º–∞—Ç –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω")
        
        # 7. –§–∏–Ω–∞–ª—å–Ω—ã–µ –ø—Ä–æ–≤–µ—Ä–∫–∏
        print(f"\n‚úÖ –§–ò–ù–ê–õ–¨–ù–´–ï –ü–†–û–í–ï–†–ö–ò:")
        
        checks = {
            "–¢–µ–∫—Å—Ç–æ–≤—ã–µ —á–∞–Ω–∫–∏ –ø–æ–ª—É—á–µ–Ω—ã": len(text_chunks_received) > 0,
            "–ê—É–¥–∏–æ —á–∞–Ω–∫–∏ –ø–æ–ª—É—á–µ–Ω—ã": len(audio_chunks_received) > 0,
            "–ê—É–¥–∏–æ –¥–∞–Ω–Ω—ã–µ –Ω–µ –ø—É—Å—Ç—ã–µ": total_audio_bytes > 0,
            "gRPC –æ—Ç–≤–µ—Ç—ã —Å—Ñ–æ—Ä–º–∏—Ä–æ–≤–∞–Ω—ã": len(grpc_responses) > 0,
            "–§–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS": (
                audio_info.get('channels') == 1 and
                audio_info.get('bits_per_sample') == 16 and
                audio_info.get('format') in ['pcm', 'int16']
            ),
        }
        
        all_passed = True
        for check_name, check_result in checks.items():
            status = "‚úÖ" if check_result else "‚ùå"
            print(f"   {status} {check_name}")
            if not check_result:
                all_passed = False
        
        if all_passed:
            print("\n" + "="*80)
            print("üéâ –í–°–ï –ü–†–û–í–ï–†–ö–ò –ü–†–û–ô–î–ï–ù–´!")
            print("="*80)
            print("\n‚úÖ –ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–æ:")
            print("   1. gRPC –∑–∞–ø—Ä–æ—Å –æ–±—Ä–∞–±–æ—Ç–∞–Ω –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ")
            print("   2. LLM –æ—Ç–≤–µ—Ç –ø–æ–ª—É—á–µ–Ω –∏ –æ–±—Ä–∞–±–æ—Ç–∞–Ω")
            print("   3. Edge TTS —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–ª –∞—É–¥–∏–æ –≤ —Ñ–æ—Ä–º–∞—Ç–µ Azure TTS")
            print("   4. –ê—É–¥–∏–æ —á–∞–Ω–∫–∏ –ø–µ—Ä–µ–¥–∞–Ω—ã —á–µ—Ä–µ–∑ gRPC –∫–ª–∏–µ–Ω—Ç—É")
            print("   5. –§–æ—Ä–º–∞—Ç –∞—É–¥–∏–æ —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS (24kHz, 16-bit, mono, PCM)")
            print("   6. gRPC –æ—Ç–≤–µ—Ç—ã —Å—Ñ–æ—Ä–º–∏—Ä–æ–≤–∞–Ω—ã –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ (dtype='int16')")
            return True
        else:
            print("\n" + "="*80)
            print("‚ùå –ù–ï–ö–û–¢–û–†–´–ï –ü–†–û–í–ï–†–ö–ò –ü–†–û–í–ê–õ–ï–ù–´")
            print("="*80)
            return False
        
    except Exception as e:
        print(f"\n‚ùå –û–®–ò–ë–ö–ê: {e}")
        import traceback
        traceback.print_exc()
        return False
    finally:
        # Cleanup
        if 'audio_processor' in locals():
            await audio_processor.cleanup()
        if 'workflow' in locals():
            # Cleanup workflow –µ—Å–ª–∏ –Ω—É–∂–Ω–æ
            pass


async def test_audio_format_compatibility():
    """
    –î–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω—ã–π —Ç–µ—Å—Ç: –ø—Ä–æ–≤–µ—Ä–∫–∞ —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç–∏ —Ñ–æ—Ä–º–∞—Ç–∞ –∞—É–¥–∏–æ —Å Azure TTS
    """
    print("\n" + "="*80)
    print("–¢–ï–°–¢ –°–û–í–ú–ï–°–¢–ò–ú–û–°–¢–ò –§–û–†–ú–ê–¢–ê –ê–£–î–ò–û")
    print("="*80)
    
    try:
        from modules.audio_generation.core.audio_processor import AudioProcessor
        
        print("\n[1/3] –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è AudioProcessor...")
        audio_processor = AudioProcessor()
        await audio_processor.initialize()
        print("   ‚úÖ AudioProcessor –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        
        print("\n[2/3] –ì–µ–Ω–µ—Ä–∞—Ü–∏—è —Ç–µ—Å—Ç–æ–≤–æ–≥–æ –∞—É–¥–∏–æ...")
        test_text = "This is a test audio generation to verify format compatibility."
        
        chunks = []
        total_bytes = 0
        
        async for audio_chunk in audio_processor.generate_speech_streaming(test_text):
            chunks.append(audio_chunk)
            total_bytes += len(audio_chunk)
        
        print(f"   ‚úÖ –ê—É–¥–∏–æ —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω–æ: {len(chunks)} —á–∞–Ω–∫–æ–≤, {total_bytes} –±–∞–π—Ç")
        
        print("\n[3/3] –ü—Ä–æ–≤–µ—Ä–∫–∞ —Ñ–æ—Ä–º–∞—Ç–∞...")
        audio_info = audio_processor.get_audio_info()
        
        format_checks = {
            "Channels = 1 (mono)": audio_info.get('channels') == 1,
            "Bits per sample = 16": audio_info.get('bits_per_sample') == 16,
            "Format = PCM/int16": audio_info.get('format') in ['pcm', 'int16'],
            "Chunks divisible by 2 (16-bit)": all(len(c) % 2 == 0 for c in chunks),
            "Total bytes > 0": total_bytes > 0,
            f"Sample rate = {audio_info.get('sample_rate')} Hz": audio_info.get('sample_rate') in [24000, 48000],
        }
        
        all_passed = True
        for check_name, check_result in format_checks.items():
            status = "‚úÖ" if check_result else "‚ùå"
            print(f"   {status} {check_name}")
            if not check_result:
                all_passed = False
        
        if all_passed:
            print("\n   ‚úÖ –§–æ—Ä–º–∞—Ç –ø–æ–ª–Ω–æ—Å—Ç—å—é —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS!")
            return True
        else:
            print("\n   ‚ùå –§–æ—Ä–º–∞—Ç –Ω–µ —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS")
            return False
        
    except Exception as e:
        print(f"\n‚ùå –û–®–ò–ë–ö–ê: {e}")
        import traceback
        traceback.print_exc()
        return False
    finally:
        if 'audio_processor' in locals():
            await audio_processor.cleanup()


async def main():
    """–ó–∞–ø—É—Å–∫ –≤—Å–µ—Ö —Ç–µ—Å—Ç–æ–≤"""
    print("="*80)
    print("–ü–û–õ–ù–û–ï E2E –¢–ï–°–¢–ò–†–û–í–ê–ù–ò–ï –ü–ê–ô–ü–õ–ê–ô–ù–ê")
    print("="*80)
    print("\n–¢–µ—Å—Ç–∏—Ä—É–µ–º –≤–µ—Å—å –ø—Ä–æ—Ü–µ—Å—Å –æ—Ç –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏ –æ—Ç–≤–µ—Ç–∞ –∞—Å—Å–∏—Å—Ç–µ–Ω—Ç–∞")
    print("–¥–æ –ø–µ—Ä–µ–¥–∞—á–∏ –∞—É–¥–∏–æ –Ω–∞ –∫–ª–∏–µ–Ω—Ç—Å–∫—É—é —á–∞—Å—Ç—å —á–µ—Ä–µ–∑ gRPC")
    
    results = {}
    
    # –¢–µ—Å—Ç 1: –ü–æ–ª–Ω—ã–π –ø–∞–π–ø–ª–∞–π–Ω
    print("\n" + "="*80)
    print("–¢–ï–°–¢ 1: –ü–û–õ–ù–´–ô –ü–ê–ô–ü–õ–ê–ô–ù E2E")
    print("="*80)
    results['full_pipeline'] = await test_full_pipeline_e2e()
    
    # –¢–µ—Å—Ç 2: –°–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç—å —Ñ–æ—Ä–º–∞—Ç–∞
    print("\n" + "="*80)
    print("–¢–ï–°–¢ 2: –°–û–í–ú–ï–°–¢–ò–ú–û–°–¢–¨ –§–û–†–ú–ê–¢–ê –ê–£–î–ò–û")
    print("="*80)
    results['format_compatibility'] = await test_audio_format_compatibility()
    
    # –ò—Ç–æ–≥–æ–≤—ã–π –æ—Ç—á–µ—Ç
    print("\n" + "="*80)
    print("–ò–¢–û–ì–û–í–´–ô –û–¢–ß–ï–¢")
    print("="*80)
    
    for test_name, success in results.items():
        status = "‚úÖ –ü–†–û–ô–î–ï–ù" if success else "‚ùå –ü–†–û–í–ê–õ–ï–ù"
        print(f"   {status}: {test_name}")
    
    total_tests = len(results)
    passed_tests = sum(1 for s in results.values() if s)
    
    print(f"\nüìä –°—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞:")
    print(f"   –í—Å–µ–≥–æ —Ç–µ—Å—Ç–æ–≤: {total_tests}")
    print(f"   –ü—Ä–æ–π–¥–µ–Ω–æ: {passed_tests}")
    print(f"   –ü—Ä–æ–≤–∞–ª–µ–Ω–æ: {total_tests - passed_tests}")
    print(f"   –£—Å–ø–µ—à–Ω–æ—Å—Ç—å: {passed_tests * 100 // total_tests if total_tests > 0 else 0}%")
    
    if passed_tests == total_tests:
        print("\nüéâ –í–°–ï –¢–ï–°–¢–´ –ü–†–û–ô–î–ï–ù–´!")
        print("\n‚úÖ –ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–æ:")
        print("   ‚Ä¢ –ü–æ–ª–Ω—ã–π –ø–∞–π–ø–ª–∞–π–Ω —Ä–∞–±–æ—Ç–∞–µ—Ç –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ")
        print("   ‚Ä¢ Edge TTS –≥–µ–Ω–µ—Ä–∏—Ä—É–µ—Ç –∞—É–¥–∏–æ –≤ —Ñ–æ—Ä–º–∞—Ç–µ Azure TTS")
        print("   ‚Ä¢ –ê—É–¥–∏–æ –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ –ø–µ—Ä–µ–¥–∞–µ—Ç—Å—è —á–µ—Ä–µ–∑ gRPC –∫–ª–∏–µ–Ω—Ç—É")
        print("   ‚Ä¢ –§–æ—Ä–º–∞—Ç —Å–æ–≤–º–µ—Å—Ç–∏–º —Å Azure TTS (24kHz, 16-bit, mono, PCM)")
        return 0
    else:
        print("\n‚ö†Ô∏è  –ù–ï–ö–û–¢–û–†–´–ï –¢–ï–°–¢–´ –ü–†–û–í–ê–õ–ï–ù–´")
        return 1


if __name__ == "__main__":
    exit_code = asyncio.run(main())
    sys.exit(exit_code)
